File: Data Structures & Algorithms/two-integer-sum-ii/submission-1.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int i = 0;
        int j = numbers.length - 1;

        while (i < j){
            int currentSum = numbers[i] + numbers[j];
            if (currentSum == target){
                return new int[]{i + 1, j + 1};
            }

            if (currentSum > target){
                j--;
            } else {
                i++;
            }
        }
        return new int[]{};
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:

        # If the target > than 1st number of a row, then target COULD be inside the row ONLY if 
        # target is smaller than the last number of the row

        # we know that if the target is smaller than the 1st number of the row, 
        # then target must live inside the previous row

        top, bot = 0, len(matrix) - 1

        while top <= bot:
            mid = (top + bot ) // 2
            first_ele = matrix[mid][0]

            if first_ele > target:
                bot = mid - 1
            elif first_ele <= target:
                if first_ele == target:
                    return True
                
                l, r = 0, len(matrix[mid]) - 1
                while l <= r:
                    m = (l+r) // 2
                    element = matrix[mid][m]
                    if element < target:
                        l = m + 1
                    elif element > target:
                        r = m - 1
                    else:
                        return True
                top = mid + 1

        return False